cfy_cluster_manager/main.py

```diff
@@ -962,15 +968,74 @@ def remove(config_path, verbose):
                       if using_three_nodes_cluster else
                       _generate_general_cluster_dict(config))
 
-    previous_installation = _previous_installation(instances_dict)
-    if previous_installation:
+    if _previous_installation(instances_dict):
         _handle_installed_instances(instances_dict, True, verbose)
-        _print_successful_installation_message(start_time, 'removed')
+        _print_success_message(start_time, 'removed')
     else:
         logger.info('No previous installation of a Cloudify cluster was '
                     'detected. Nothing to remove.')
 
 
+def _upgrade_cluster(instances_dict, verbose, upgrade_rpm_path):
+    for instance_type, instances_list in instances_dict.items():
+        for instance in instances_list:
+            logger.info('Upgrading %s', instance.name)
+            if '5.1.0' in instance.get_version():
+                logger.info('Installing new RPM on %s', instance.name)
+                instance.run_command(
+                    'yum install -y {rpm}'.format(rpm=upgrade_rpm_path),
+                    use_sudo=True, hide_stdout=True)
+            logger.info('Running upgrade command on %s', instance.name)
+            instance.run_command(
+                'cfy_manager upgrade -c {config} {verbose}'.format(
+                    config=instance.config_path,
+                    verbose='-v' if verbose else '')
+            )
+
+
+def _should_upgrade(instances_dict, using_three_nodes_cluster):
+    logger.info('Verifying v5.1.0 is installed on all instances')
+    for instance_type, instances_list in instances_dict.items():
+        for instance in instances_list:
+            logger.debug('Verifying instance %s', instance.private_ip)
+            if not _rpm_was_installed(instance):
+                logger.debug('cloudify-manager-install is not installed on %s',
+                             instance.private_ip)
+                return False
+            instance_version = instance.get_version()
+            if '5.1.0' not in instance_version:
```

Comment:
What about comparing the upgraded version with current version and if the current version is less than the upgraded version then we can move forward, with that we can avoid hardcodeded any version where we need to do a code update on each release

Reply:
That's a great idea. Please refer to my first comment, that we don't know yet what would be the upgrade procedure for future releases.